components/proxy/src/main/java/com/hotels/styx/startup/StyxPipelineFactory.java

```diff
@@ -0,0 +1,143 @@
+import com.hotels.styx.Environment;
+import com.hotels.styx.StyxConfig;
+import com.hotels.styx.api.HttpHandler2;
+import com.hotels.styx.api.HttpInterceptor;
+import com.hotels.styx.api.service.spi.StyxService;
+import com.hotels.styx.client.applications.BackendService;
+import com.hotels.styx.infrastructure.Registry;
+import com.hotels.styx.proxy.StyxBackendServiceClientFactory;
+import com.hotels.styx.proxy.interceptors.ConfigurationContextResolverInterceptor;
+import com.hotels.styx.proxy.interceptors.HopByHopHeadersRemovingInterceptor;
+import com.hotels.styx.proxy.interceptors.HttpMessageLoggingInterceptor;
+import com.hotels.styx.proxy.interceptors.RequestEnrichingInterceptor;
+import com.hotels.styx.proxy.interceptors.UnexpectedRequestContentLengthRemover;
+import com.hotels.styx.proxy.interceptors.ViaHeaderAppendingInterceptor;
+import com.hotels.styx.proxy.plugin.NamedPlugin;
+import com.hotels.styx.routing.HttpPipelineFactory;
+import com.hotels.styx.routing.StaticPipelineFactory;
+import com.hotels.styx.routing.config.BuiltinInterceptorsFactory;
+import com.hotels.styx.routing.config.HttpHandlerFactory;
+import com.hotels.styx.routing.config.RouteHandlerDefinition;
+import com.hotels.styx.routing.config.RouteHandlerFactory;
+import com.hotels.styx.routing.handlers.BackendServiceProxy;
+import com.hotels.styx.routing.handlers.ConditionRouter;
+import com.hotels.styx.routing.handlers.HttpInterceptorPipeline;
+import com.hotels.styx.routing.handlers.ProxyToBackend;
+import com.hotels.styx.routing.handlers.StaticResponseHandler;
+import com.hotels.styx.routing.interceptors.RewriteInterceptor;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.hotels.styx.api.configuration.ConfigurationContextResolver.EMPTY_CONFIGURATION_CONTEXT_RESOLVER;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Produces the pipeline for the Styx proxy server.
+ */
+public final class StyxPipelineFactory {
+    private StyxPipelineFactory() {
+    }
+
+    public static HttpHandler2 create(CoreConfig config) {
+        BuiltinInterceptorsFactory builtinInterceptorsFactory = new BuiltinInterceptorsFactory(
+                ImmutableMap.of("Rewrite", new RewriteInterceptor.ConfigFactory()));
+
+        Map<String, HttpHandlerFactory> objectFactories = createBuiltinRoutingObjectFactories(
+                config.environment(),
+                config.services(),
+                config.plugins(),
+                builtinInterceptorsFactory);
+
+        RouteHandlerFactory routeHandlerFactory = new RouteHandlerFactory(objectFactories, new ConcurrentHashMap<>());
+
+        return styxHttpPipeline(
+                config.environment().styxConfig(),
+                configuredPipeline(config.environment(), config.services(), config.plugins(), routeHandlerFactory));
+    }
+
+    private static HttpHandler2 styxHttpPipeline(StyxConfig config, HttpHandler2 interceptorsPipeline) {
```

Comment:
This "unit" of code adds the default "built-in" interceptors around the configured HTTP pipeline. So we should have following unit tests:

 - Adds default interceptors in front of user-configured pipeline.
 - Adds logging interceptor when logging configuration is enabled
 - Enables long logging messages when ...

Reply:
However, in order to make this easier, the `StandardHttpPipeline` class should be amalgamated into `HttpInterceptorPipeline`. Then you could expose its `interceptors` list as a `@VisibleForTesting` attribute, and ensure the interceptors are prepended as expected.

Reply:
This code was copied wholesale from the previous version of `StyxServer`, so I will take some time to wrap my head around all of it. It may have to be done as a separate increment - I don't know yet.